chalk-solve/src/coherence/solve.rs

```diff
@@ -131,7 +130,7 @@ impl<I: Interner, S: Solver<I>, SC: Into<S> + Copy> CoherenceSolver<'_, I, S, SC
             .negate(interner);
 
         let canonical_goal = &goal.into_closed_goal(interner);
-        let solution = self.solver_choice.into().solve(self.db, canonical_goal);
+        let solution = self.solver.solve(self.db, canonical_goal);
```

Comment:
So...this is not the same things semantically. Before, we use a fresh solver for each goal. Now, we use a single solver for all goals. Importantly, the solver can (and does) keep some state between goals. It's a bit surprising to me that nothing fails here. (Really, I think the only thing we have to be careful about is placeholders). I think this needs to stay as creating a new solver for each goal, but with some convincing (from myself or otherwise), this could be changed now or later.